timemachine/fe/interpolate.py

```diff
@@ -0,0 +1,217 @@
+import numpy as np
+
+
+def _add_dst_to_src_bond_or_angle(src_idxs, src_params, dst_idxs, dst_params):
+    """
+    eg:
+    src_idxs: [[4,9], [3,4]]
+    src_params: [[a, b], [c, d]]
+
+    dst_idxs: [[3,4], [9,5]]
+    dst_params: [[e, f], [g, h]]
+
+    zero_flags: [True, False]
+
+    Generate end-state key value pairs
+
+    src_idxs: [[4,9], [3,4], [9,5]]
+    src_params: [[a,b], [c,d], [0,h]
+
+    """
+    # used only for bond and angle terms
+    #
+    # note that *missing* terms are the result of core-hopping, or chiral
+    # inversions on core atoms. dummy atom interactions (bonds/angles) are fully maintained
+    # at the end-state.
+    assert len(set((src_idxs))) == len(src_idxs)
+    assert len(set((dst_idxs))) == len(dst_idxs)
+
+    src_kv = dict(zip(src_idxs, src_params))
+    for idxs, params in zip(dst_idxs, dst_params):
+        assert idxs[0] < idxs[-1]
+        if idxs not in src_kv:
+            # new_params = []
+            _, bond_or_angle = params
+            src_kv[idxs] = (0, bond_or_angle)
+
+    return src_kv
+
+
+def align_harmonic_bond_or_angle_idxs_and_params(src_idxs, src_params, dst_idxs, dst_params):
+
+    # sanitize
+    src_idxs = [tuple(p) for p in src_idxs]
+    src_params = [tuple(p) for p in src_params]
+    dst_idxs = [tuple(p) for p in dst_idxs]
+    dst_params = [tuple(p) for p in dst_params]
+
+    src_kv = _add_dst_to_src_bond_or_angle(src_idxs, src_params, dst_idxs, dst_params)
+    dst_kv = _add_dst_to_src_bond_or_angle(dst_idxs, dst_params, src_idxs, src_params)
+    assert src_kv.keys() == dst_kv.keys()
+
+    res = set()
+
+    for k in src_kv.keys():
+        res.add((tuple(k), src_kv[k], dst_kv[k]))
+
+    return res
```

Comment:
nit: I think this could be simplified a bit (and remove the need for the helper function) as something like:

```python
def align_harmonic_bond_or_angle_idxs_and_params(src_idxs, src_params, dst_idxs, dst_params):

    src_kv = dict(zip(src_idxs, src_params))
    dst_kv = dict(zip(dst_idxs, dst_params))

    aligned = {
        (
            k,
            tuple(src_kv[k]) if k in src_kv else (0, dst_kv[k][1]),
            tuple(dst_kv[k]) if k in dst_kv else (0, src_kv[k][1]),
        )
        for k in set(src_kv.keys()).union(dst_kv)
    }

    return aligned
```

Reply:
I think the various `align_*` functions defined here have a common pattern that might be useful to factor out.

E.g. with a common helper like
```python
def align_idxs_and_params(
    src_idxs,
    src_params,
    dst_idxs,
    dst_params,
    make_default,
    key=lambda idxs, _: idxs,
    get_idxs=lambda key: key,
):
    def to_hashable(x):
        return x if isinstance(x, Hashable) else tuple(x)

    src_kv = {to_hashable(key(idxs, params)): to_hashable(params) for idxs, params in zip(src_idxs, src_params)}
    dst_kv = {to_hashable(key(idxs, params)): to_hashable(params) for idxs, params in zip(dst_idxs, dst_params)}

    return {
        (
            get_idxs(k),
            src_kv[k] if k in src_kv else make_default(dst_kv[k]),
            dst_kv[k] if k in dst_kv else make_default(src_kv[k]),
        )
        for k in set(src_kv.keys()).union(dst_kv)
    }
```,

then the others can become one-liners, e.g.

```python
align_harmonic_bond_or_angle_idxs_and_params = partial(align_idxs_and_params, make_default=lambda p: (0, p[1]))
align_nonbonded_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: (0, 0, 0))
align_chiral_atom_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: 0)

align_torsion_idxs_and_params = partial(
    align_idxs_and_params,
    make_default=lambda p: (0, p[1], p[2]),
    key=lambda idxs, params: (idxs, params[2]),
    get_idxs=lambda key: key[0],
)
```